lib/pavilion/lockfile.py

```diff
@@ -0,0 +1,191 @@
+from __future__ import division, unicode_literals, print_function
+
+import grp
+import logging
+import os
+import time
+import uuid
+
+LOGGER = logging.getLogger('pav.' + __name__)
+
+
+# Expires after a silly long time.
+NEVER = 10**10
+
+
+class TimeoutError(RuntimeError):
+    """Error raised when the lockfile times out."""
+    pass
+
+
+class LockFile(object):
+    """An NFS friendly way to create a lock file. Locks contain information on what host and user
+    created the lock, and have a built in expiration date. To be used in a 'with' context.
+    :cvar DEFAULT_EXPIRE: How long it takes for a lock to expire."""
+
+    # Time till file is considered stale, in seconds. (5 minute default)
+    DEFAULT_EXPIRE = 60 * 60 * 5
+
+    # How long to sleep between lock attempts.
+    # This shouldn't be any less than 0.01 or so on a regular filesystem. 0.2 is pretty reasonable
+    # for an nfs filesystem and sporadically used locks.
+    SLEEP_PERIOD = 0.2
+
+    # Default lock permissions
+    LOCK_PERMS = 0774
+
+    def __init__(self, lockfile_path, group=None, timeout=None, expires_after=DEFAULT_EXPIRE):
+        """Initialize the lock file. The resulting class can be reused multiple times.
+        :param unicode lockfile_path: The path to the lockfile. Should probably start with a '.',
+        and end with '.lock', but that's up to the user.
+        :param unicode group: The name of the group to set lockfiles to. If this is given,
+        :param timeout: When to quit trying to acquire the lock in seconds. None (default) denotes
+        non-blocking mode.
+        :param expires_after: When to consider the lock dead, and overwritable (in seconds). The
+        NEVER module variable is provided as easily named long time. (10^10 secs, 317 years)
+        """
+
+        self._lock_path = lockfile_path
+        self._timeout = timeout
+        self._expire_period = expires_after
+        self._group = None
+
+        if group is not None:
+            # This could error out, but should be checked by pavilion separately.
+            self._group = grp.getgrnam(group).gr_gid
```

Comment:
In my script-composer stuff, I still wrapped this in a try:except block and had it raise a ValueError.  Other parts of the code could handle it, but they can just as easily handle an exception being raised from the initializer call to this function.

Reply:
The assumption is that this will always be the group specified in Pavilion's configuration, which should be checked for validity in advance. You're right; it's a bad assumption.